Sources/NIOSOCKS/Messages/ServerResponse.swift

```diff
@@ -0,0 +1,114 @@
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+// MARK: - ServerResponse
+
+/// The SOCKS Server's response to the client's request
+/// indicating if the request succeeded or failed.
+struct ServerResponse: Hashable {
+    
+    /// The SOCKS protocol version - we currently only support v5.
+    public let version: UInt8 = 5
+    
+    /// The status of the connection - used to check if the request
+    /// succeeded or failed.
+    public var reply: Reply
+    
+    /// The host address.
+    public var boundAddress: AddressType
+    
+    /// Creates a new `ServerResponse`.
+    /// - parameter reply: The status of the connection - used to check if the request
+    /// succeeded or failed.
+    /// - parameter boundAddress: The host address.
+    public init(reply: Reply, boundAddress: AddressType) {
+        self.reply = reply
+        self.boundAddress = boundAddress
+    }
+}
+
+extension ByteBuffer {
+    
+    mutating func readServerResponse() throws -> ServerResponse? {
+        let save = self
+        guard
+            let version = self.readInteger(as: UInt8.self),
+            let reply = self.readInteger(as: UInt8.self).flatMap({ Reply(value: $0) }),
+            let reserved = self.readInteger(as: UInt8.self),
+            let boundAddress = try self.readAddresType()
+        else {
+            self = save
+            return nil
+        }
+        
+        guard reserved == 0x0 else {
+            self = save
+            throw InvalidReservedByte(actual: reserved)
+        }
+        
+        guard version == 0x05 else {
+            self = save
+            throw InvalidProtocolVersion(actual: version)
+        }
+        
+        return .init(reply: reply, boundAddress: boundAddress)
+    }
+    
+}
+
+// MARK: - Reply
+
+/// Used to indicate if the SOCKS client's connection request succeeded
+/// or failed.
+public struct Reply: Hashable {
+    
+    /// The connection succeeded and data can now be transmitted.
+    static let succeeded = Reply(value: 0x00)
+    
+    /// The SOCKS server encountered an internal failure.
+    static let serverFailure = Reply(value: 0x01)
+    
+    /// The connection to the host was not allowed.
+    static let notAllowed = Reply(value: 0x02)
+    
+    /// The host network is not reachable.
+    static let networkUnreachable = Reply(value: 0x03)
+    
+    /// The target host was not reachable.
+    static let hostUnreachable = Reply(value: 0x04)
+    
+    /// The connection tot he host was refused
+    static let refused = Reply(value: 0x05)
+    
+    /// The host address's TTL has expired.
+    static let ttlExpired = Reply(value: 0x06)
+    
+    /// The provided command is not supported.
+    static let commandUnsupported = Reply(value: 0x07)
+    
+    /// The provided address type is not supported.
+    static let addressUnsupported = Reply(value: 0x08)
+    
+    /// The raw `UInt8` status code.
+    public let value: UInt8
```

Comment:
Nit: no reason to prevent mutation of this.

```suggestion
    public var value: UInt8
```